core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4QueryHttpSigner.java

```diff
@@ -15,27 +15,81 @@
+import static software.amazon.awssdk.http.auth.internal.util.SignerUtils.validatedProperty;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
 import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.SdkHttpRequest;
 import software.amazon.awssdk.http.auth.AwsV4QueryHttpSigner;
-import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
-import software.amazon.awssdk.http.auth.spi.AsyncSignedRequest;
-import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
-import software.amazon.awssdk.http.auth.spi.SyncSignedRequest;
+import software.amazon.awssdk.http.auth.internal.checksums.ContentChecksum;
+import software.amazon.awssdk.http.auth.internal.util.CanonicalRequestV2;
+import software.amazon.awssdk.http.auth.internal.util.CredentialScope;
+import software.amazon.awssdk.http.auth.internal.util.SignerConstant;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
 import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.identity.spi.AwsSessionCredentialsIdentity;
+import software.amazon.awssdk.utils.Pair;
 
 /**
  * A default implementation of {@link AwsV4QueryHttpSigner}.
  */
 @SdkInternalApi
-public final class DefaultAwsV4QueryHttpSigner implements AwsV4QueryHttpSigner {
+public final class DefaultAwsV4QueryHttpSigner extends DefaultAwsV4HttpSigner implements AwsV4QueryHttpSigner {
 
     @Override
-    public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity> request) {
-        throw new UnsupportedOperationException();
+    public void addSessionCredentials(SdkHttpRequest.Builder requestBuilder,
+                                      AwsSessionCredentialsIdentity credentials) {
+        requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_SECURITY_TOKEN, credentials.sessionToken());
+    }
+
+    @Override
+    public void addPrerequisites(SdkHttpRequest.Builder requestBuilder,
+                                 SignRequest<?, ? extends AwsCredentialsIdentity> signRequest,
+                                 ContentChecksum contentChecksum) {
+        Instant requestSigningInstant = validatedProperty(signRequest, SIGNING_CLOCK).instant();
+        String regionName = validatedProperty(signRequest, REGION_NAME);
+        String serviceSigningName = validatedProperty(signRequest, SERVICE_SIGNING_NAME);
+        Duration expirationDuration = validateExpirationDuration(validatedProperty(signRequest, EXPIRATION_DURATION,
+            PRESIGN_URL_MAX_EXPIRATION_DURATION));
+
+        AwsCredentialsIdentity credentials = sanitizeCredentials(signRequest.identity());
+        CredentialScope credentialScope = new CredentialScope(regionName, serviceSigningName, requestSigningInstant);
+
+        addHostHeader(requestBuilder);
+
+        List<Pair<String, List<String>>> canonicalHeaders = getCanonicalHeaders(requestBuilder.build().headers());
```

Comment:
Making the caller get the canonical headers before asking for signed headers `getSignedHeadersString(getCanonicalHeaders())` seems more to ask for the caller, risks bug at call site forgetting to call getCanonicalHeaders. Outside of `getSignedHeadersString()` (overloaded for an already created immutable CanonicalRequest v/s statically for the query case) I don't see the need for CanonicalRequestV2 to have any other non-private method than `getString()`.

Reply:
You'd have to try pretty hard to give the wrong headers to `getSignedHeadersString(List<Pair<String, List<String>>> canonicalHeaders)` - I get the spirit of your comment, but I don't necessarily agree in this case.

All the members are set at object construction, so calling `getSignedHeadersString()` is guaranteed to get you the string. 

Based on your other comments about keeping things private before they are needed, sure, I can do that.